File: 프로그래머스/lv2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    ans = 0

    # 모든 던전 순열에 대해 탐색
    for dungeon_perm in permutations(dungeons, len(dungeons)):
        hp = k
        count = 0

        # 순열의 각 던전을 방문하며 체력을 소모함
        for dungeon in dungeon_perm:
            if hp >= dungeon[0]:
                hp -= dungeon[1]
                count += 1
            else:
                break

        # 최대 클리어한 던전 수를 저장함
        ans = max(ans, count)

    return ans

File: 프로그래머스/lv2/87946. 피로도/피로도.js
function solution(k, dungeons) {
  let visited = Array(dungeons.length).fill(false);
  let ans = 0;

  let dfs = (hp, count, visited) => {
    ans = Math.max(ans, count); // 현재 count와 ans 중 최대 값을 ans에 저장합니다.

    for (let i = 0; i < dungeons.length; i++) {
      if (visited[i] || dungeons[i][0] > hp) {
        continue;
      }

      visited[i] = true;
      dfs(hp - dungeons[i][1], count + 1, visited);
      visited[i] = false;
    }
  };

  dfs(k, 0, visited); // 초기 호출

  return ans; // 계산된 정답 반환
}